tests/tests.rs

```diff
@@ -1,99 +1,124 @@
+            #[test] fn $pass() { _test_pass(&mut Vec::new(), vec![$data; 1024]); }
+            #[test] fn $fail() { _test_fail(&mut Vec::new(), vec![$data; 1024]); }
+            #[test] fn $size() { _test_size(&mut Vec::new(), vec![$data; 1024]); }
+        )*
+    };
+}
+gen_tests!{
+    [4, 1, 2] => (test_array_pass,
+                  test_array_fail,
+                  test_array_size),
+
+    std::num::NonZeroI32::new(1) => (test_nonzero_pass,
+                                     test_nonzero_fail,
+                                     test_nonzero_size),
+
+    Some(vec![8, 1, 2]) => (test_option_vec_pass,
+                            test_option_vec_fail,
+                            test_option_vec_size),
+
+    (format!("x"), vec![1, 2, 3])  => (test_alignment_pass,
+                                       test_alignment_fail,
+                                       test_alignment_size),
+
+    (format!("x"), vec![1u128, 2, 3]) => (test_alignment_128_pass,
+                                          test_alignment_128_fail,
+                                          test_alignment_128_size),
+
+    Some(Box::new(9u64)) => (test_option_box_u64_pass,
+                             test_option_box_u64_fail,
+                             test_option_box_u64_size),
+
+    (3u32, 8u32, 1u32, 7u32) => (test_u32x4_pass,
+                                 test_u32x4_fail,
+                                 test_u32x4_size),
+
+    42u64 => (test_u64_pass,
+              test_u64_fail,
+              test_u64_size),
+
+    687u128 => (test_u128_pass,
+                test_u128_fail,
+                test_u128_size),
+
+    format!("grawwwwrr!") => (test_string_pass,
+                              test_string_fail,
+                              test_string_size),
+
+    vec![(0u64, format!("grawwwwrr!")); 32] => (test_vec_u_s_pass,
+                                                test_vec_u_s_fail,
+                                                test_vec_u_s_size),
+
+    MyStruct{ a: "test".to_owned(),
+              b: 0,
+              c: vec![0, 1, 2],
+              d: &"grawwwwrr!" } => (test_macro_pass,
+                                     test_macro_fail,
+                                     test_macro_size),
+
+    PhantomNotAbo::default() => (test_phantom_notabo_pass,
+                                 test_phantom_notabo_fail,
+                                 test_phantom_notabo_size),
+
+    Some(&42u64) => (test_ref_u64_pass,
+                     test_ref_u64_fail,
+                     test_ref_u64_size),
+
+    &"grawwwwrr!" => (test_str_pass,
+                      test_str_fail,
+                      test_str_size),
+
+    &[0, 1, 2] => (test_slice_pass,
+                   test_slice_fail,
+                   test_slice_size),
 }
 
-fn _test_pass<T: Abomonation+Eq>(record: T) {
-    let mut bytes = Vec::new();
+fn _test_pass<'bytes, T: Abomonation<'bytes> + Debug + Eq>(
+    mut bytes: &'bytes mut Vec<u8>, record: T
+) {
```

Comment:
These changes work around limitations of Rust's current API vocabulary with respect to lifetimes.

Basically, there is currently no way to express "we want T to implement `Abomonation<'bytes>` for all possibly `'bytes` lifetimes on this function's stack". We only have "it should work for all slices of bytes" (which excludes types containing references) and "it should work for whatever byte container was passed as a parameter to the function" (what I ended up using).

Amusingly enough, this is closely related, but not quite identical, to the reason why a full port of `Abomonated` to the new API is not possible.